File: code/shared/types/api-response.ts
/**
 * Standard API response types shared between backend and frontend.
 */

export interface ApiResponse<T = unknown> {
  success: true;
  data: T;
  timestamp: string;
}

export interface ApiErrorResponse {
  success: false;
  error: {
    code: string;
    message: string;
    details?: Record<string, unknown>;
  };
  timestamp: string;
}

export interface PaginatedResponse<T = unknown> {
  success: true;
  data: T[];
  meta: {
    total: number;
    page: number;
    limit: number;
    totalPages: number;
  };
  timestamp: string;
}

export type UserRole =
  | 'IT_ADMIN'
  | 'LANH_DAO'
  | 'CAN_BO_QUAN_LY'
  | 'CAN_BO_CO_SO'
  | 'SUBJECT';

export type DataScopeLevel = 'WARD' | 'DISTRICT' | 'PROVINCE' | 'SYSTEM';

export type AlertLevel = 'THAP' | 'TRUNG_BINH' | 'CAO' | 'KHAN_CAP';

export type UserStatus = 'ACTIVE' | 'LOCKED' | 'DEACTIVATED';

File: code/mobile/app/src/main/java/com/example/smtts/data/model/ApiResponse.kt
package com.example.smtts.data.model

import com.google.gson.annotations.SerializedName

/**
 * Generic API response wrapper matching backend format:
 * { "success": true, "data": T, "timestamp": "..." }
 */
data class ApiResponse<T>(
    @SerializedName("success") val success: Boolean,
    @SerializedName("data") val data: T,
    @SerializedName("timestamp") val timestamp: String
)

/**
 * API error response format:
 * { "success": false, "error": { "code": "...", "message": "..." }, "timestamp": "..." }
 */
data class ApiErrorResponse(
    @SerializedName("success") val success: Boolean,
    @SerializedName("error") val error: ApiErrorDetail,
    @SerializedName("timestamp") val timestamp: String
)

data class ApiErrorDetail(
    @SerializedName("code") val code: String,
    @SerializedName("message") val message: String
)
